### ISSUE TITLE
HostPath mount type FileOrCreate creates a directory on the node

### ISSUE THREAD

**Issue by LRueckert** [kubernetes/kubernetes#65825]
/kind bug

**Context**:
- E2E test `[k8s.io] KubeletManagedEtcHosts should test kubelet managed /etc/hosts file [Conformance]` failed because `busybox-3` (the one with the volume mount doesn't start)
- doesn't start due to mounting error: 
```
Failed: Error: failed to start container "busybox-3": Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "process_linux.go:402: container init caused \"rootfs_linux.go:58: mounting \\\"/tmp/etc-hosts708091275\\\" to rootfs \\\"/var/lib/docker/overlay2/a0d8a562c40788ec1aab2f5ec2283eb576c13d08b36db3a523452c69339f57a5/merged\\\" at \\\"/var/lib/docker/overlay2/a0d8a562c40788ec1aab2f5ec2283eb576c13d08b36db3a523452c69339f57a5/merged/etc/hosts\\\" caused \\\"not a directory\\\"\"": unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
```
  - copied the failing pod and reduced it to the parts that are failing

**What happened**:
- creating a pod from the following template:
```
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
  namespace: test
spec:
  containers:
  - command:
    - sleep
    - "900"
    image: gcr.io/kubernetes-e2e-test-images/netexec-amd64:1.0
    imagePullPolicy: IfNotPresent
    name: busybox-3
    resources: {}
    volumeMounts:
    - mountPath: /etc/hosts
      name: host-etc-hosts
  dnsPolicy: ClusterFirst
  nodeName: node1
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  volumes:
  - hostPath:
      path: /tmp/etc-hosts462102378
      type: FileOrCreate
    name: host-etc-hosts
```
- pod creation fails with mounting error:
```
Error: failed to start container "busybox-3": Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container pr
ocess caused "process_linux.go:402: container init caused \"rootfs_linux.go:58: mounting \\\"/tmp/etc-hosts462102378\\\" to rootfs \\\"/var/lib/docker/overlay2/877fb06f923c08d83e6f36cd3ef0e2757a1b814b6bfcecb6929aeef9bd41c45a/merged\\\"
at \\\"/var/lib/docker/overlay2/877fb06f923c08d83e6f36cd3ef0e2757a1b814b6bfcecb6929aeef9bd41c45a/merged/etc/hosts\\\" caused \\\"not a directory\\\"\"": unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if
the specified host path exists and is the expected type
```
- looking onto the node where the pod was started I can see the file `/tmp/etc-hosts462102378` was actually created as a directory
```
node1 tmp # ls -l
total 8
drwxr-xr-x. 2 root root  40 Jul  4 14:32 etc-hosts462102378
...
```

**What you expected to happen**:
- the file should actually be created as an empty file (as per documentation)

**How to reproduce it (as minimally and precisely as possible)**:
- create a pod similar to the one posted above
  - hostpath mount volume
  - type FileOrCreate

**Anything else we need to know?**:
- the E2E tests were executed from within a docker container running on a local machine

**Environment**: (of the docker container)
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"archive", BuildDate:"2018-06-28T10:42:54Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"lin
ux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3+coreos.0", GitCommit:"f1b890dbbf11abe58280b3ffe17d67749f5ae70e", GitTreeState:"clean", BuildDate:"2018-05-21T17:27:17Z", GoVersion:"go1.9.3", Compiler:"gc", Platfor
m:"linux/amd64"}
```
(I also tried it with client k8s 1.10 and go 1.9.3)
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): 
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
- Kernel (e.g. `uname -a`):
`Linux 29778d7f485a 4.9.87-linuxkit-aufs #1 SMP Wed Mar 14 15:12:16 UTC 2018 x86_64 GNU/Linux`

**Comment by msau42** [comment]
cc @dixudx

**Comment by dixudx** [comment]
@LRueckert I did not reproduce this error.

**Comment by LRueckert** [comment]
I can't give any additional information. Everything I had I provided and I am not working with it anymore.

**Comment by dims** [comment]
/close

let's reopen this when there is more info available

**Comment by SimonHeimberg** [comment]
I directory was created on my system as well.

First hostPath failed with type: File, even if the file was existing. (It reported not a file.)
Then I changed to type: FileOrCreate. This failed as well.
Then I changed the path to an unexisting one. A **directory** was created.
I deleted the directory, created a file and restarted the pod. It worked now.
When I delete the file and restart the pod, a file is created.
When I change to type: File it works as well.

When I change the path or create another volume, FileOrCreate creates a directory again.

kubernetes version (server and kubectl): v1.17.4
hardware configuration: xen
OS: Ubuntu 18.04.4
Kernel: Linux xxxx 4.15.0-88-generic #88-Ubuntu SMP Tue Feb 11 20:11:34 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
